dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1101 @@
+        public EphemeralServerBased(TemporalClient client) : base(client) { /*...*/ }
+    }
+}
+```
+
+Notes:
+
+* We call the method `ActivityEnvironment.Run` instead of `ActivityEnvironment.RunActivity` to make it clear that we
+  accept any code/lambda, it does not have to be an `[Activity` method
+* `WorkflowEnvironment` is `IAsyncDisposable` in all but the oldest .NET versions, which means it can be used like
+  `await using (var env = await WorkflowEnvironment.StartLocal()) { ...` (or just in the function without a block)
+* While it may seem unnecessary to build this now, it will be very helpful during tests of the framework itself. So we
+  should go ahead and build it
+
+## Runtime Support/Extensions
+
+.NET provides several
+[extensions to the runtime library](https://learn.microsoft.com/en-us/dotnet/standard/runtime-libraries-overview#extensions-to-the-runtime-libraries)
+which many users use when making apps. We discuss here some of those patterns and how we can support those, either by
+our existing design, or with helpers in a `Temporalio.Extensions` namespace.
+
+With DI and configuration and such, this is very analogous to Spring Boot for Java.
+
+### Serialization
+
+https://learn.microsoft.com/en-us/dotnet/standard/serialization/
+
+* We use `System.Text.Json` by default and therefore support all customization/approaches it provides
+* Our JSON payload converters may offer easy approaches to customize options to create new converters
+
+### Dependency Injection
+
+https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection
+
+* We have `Temporalio.Client.ITemporalClient` and `Temporalio.Client.ITemporalConnection` to allow DI for these impls
+* Should we also do this for `Temporalio.Client.WorkflowHandle`? Is there value in injecting that?
+* Since activity instances are instantiated by the user upon registration, the user can use DI when creating those
+  classes to, say, inject a logger
+  * We'll need an example showing this
+* When we start the workflow implementation side, we may need to revisit how DI may help users without forcing it
+
+### Configuration
+
+https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration
+
+* We support the https://learn.microsoft.com/en-us/dotnet/core/extensions/options pattern with all of our `Options`
+  classes
+  * Some of the options things like activities method references can't by loaded from a file, so we should take a look
+    at helpers for converting activity strings to activity classes reflectively
+  * We need a sample showing configuring workers via config file akin to Spring Boot
+* See the host section for how it all may come together inside a worker service
+
+### Logging
+
+https://learn.microsoft.com/en-us/dotnet/core/extensions/logging
+
+* Unfortunately there is no logging in the standard library for .NET, therefore in order to support this we'd have to
+  add a dependency on https://www.nuget.org/packages/Microsoft.Extensions.Logging
```

Comment:
Note that your libraries should add a dependency to the `.Abstractions` package, `Microsoft.Extensions.Logging.Abstractions`.
I recommend that you do go this route, rather than creating your own abstraction, however thin.

Reply:
:+1: That seems to be the consensus, so I may do this. The problem we have had in the past with dependencies like these, no matter how prevalent and stable, is people eventually complain. "Why are you making me take a dependency for logging that I don't want?"

Here's a question though: When `7.0.0` of that library drops, which version should I depend on? And when a user comes to me and says "this other library I'm using in the same app depends on the other incompatible version" what do I tell them?

Reply:
Now that .NET ships versions alternating between Long Term Service and Standard Term Service, you might find that certain companies have rules about always using an LTS version (even though .NET 6's patch lifecycle extends only 6 months past .NET 7's patch lifecycle). Anecdotally, this is especially true for banks. Given that, always adopting the latest LTS once the previous LTS is nearing EoL is probably a safe strategy. You can also multi-target and use conditionals in your project so that .NET 6 users pull in 6.0 and .NET 7 users pull in 7.0, but it's a hassle and the gain is dubious in the case of abstractions libraries which have been very stable.

Package mismatches are ok: specified versions are minimums by default, so you're generally safe picking lower versions since the end user can upgrade them in their own project as desired. Visual Studio now shows transitive dependencies in the package management UI, which helps with this.

Reply:
> Package mismatches are ok: specified versions are minimums by default, so you're generally safe picking lower versions since the end user can upgrade them in their own project as desired.

This isn't usually true with semver major version bumps. To clarify my original question just in case, I was asking about 7.x vs 6.x of that specific package, not  .NET versions.

Are you saying I should depend on https://www.nuget.org/packages/Microsoft.Extensions.Logging.Abstractions/6.0.3, not https://www.nuget.org/packages/Microsoft.Extensions.Logging.Abstractions/7.0.0? And that the latter is compatible even though they did a major bump? I haven't checked their changelog yet.

It is usually very bad for a library to add a dependency to a package whose major version moves so frequently. I'm forcing the user to override my transitive version to upgrade.